### ISSUE TITLE
Memory leaks just by including doctest.h

### ISSUE THREAD

**Issue by edmundvermeulen** [doctest/doctest#205]
## Description

We upgraded from doctest 1.2.6 to 2.2.3. Now Visual Studio reports memory leaks.

### Steps to reproduce

The following simple example code gives 3 memory leaks when run.

```cpp
#include "stdafx.h"

#define DOCTEST_CONFIG_IMPLEMENT
#include "doctest.h"

int main()
{
    int flag = _CrtSetDbgFlag(_CRTDBG_REPORT_FLAG);
    _CrtSetDbgFlag(flag | _CRTDBG_ALLOC_MEM_DF | _CRTDBG_LEAK_CHECK_DF);

    return 0;
}
```

```
Detected memory leaks!
Dumping objects ->
{175} normal block at 0x004F5620, 8 bytes long.
 Data: <  O     > EC 98 4F 00 00 00 00 00 
{163} normal block at 0x004F5578, 8 bytes long.
 Data: <| O     > 7C 8E 4F 00 00 00 00 00 
{156} normal block at 0x004F57E0, 8 bytes long.
 Data: < 5O     > AC 35 4F 00 00 00 00 00 
Object dump complete.
```

### Extra information
* doctest version: 2.2.3
* Operating System: Windows 10
* Compiler+version: Microsoft Visual Studio 2015

**Comment by onqtam** [comment]
strange... valgrind on Unix doesn't report anything.

Version 2.0+ uses a lot more things from the STL - this could be a bug in the toolchain (i've had such issues many times before where a specific version of the compiler is giving trouble).

I'll try to reproduce it and investigate but I don't know when I'll get to this.

**Comment by edmundvermeulen** [comment]
The same leaks occur in VS2019 Preview 2.

Further investigation shows that this is caused by g_infoContexts being thread_local. When DOCTEST_THREAD_LOCAL is defined to nothing then the leaks go away. Is it necessary to have it be thread_local if all the tests run from the same thread?

**Comment by onqtam** [comment]
All tests always run from the same thread but if one test case spawns multiple threads and uses asserts and logging with INFO() - then that needs to be thread-local. If you don't have such a scenario you may indeed use DOCTEST_THREAD_LOCAL.

Thanks for diagnosing this - I'm very curious as to why a thread-local global std::vector is a problem. If you have any ideas I would be grateful - otherwise I'll look into it when I have the chance... But for now version 2.3 with xml reporting is the priority.

**Comment by edmundvermeulen** [comment]
Solution: creating the static thread_local object from a getter function solves it. Like this:

```cpp
    std::vector<IContextScope*>& getInfoContexts() {
        static DOCTEST_THREAD_LOCAL std::vector<IContextScope*> g_infoContexts; // for logging with INFO()
        return g_infoContexts;
    }
```

**Comment by onqtam** [comment]
Is this solution related to the ["static initialization order ‘fiasco"](https://isocpp.org/wiki/faq/ctors#static-init-order)?

This global is used only after ```main()``` is entered so all globals (including this one) should be already initialized... Thanks for letting me know it works this way! I'm still really puzzled why this is necessary though...

If I don't come up with anything else as a solution I might add this ... "fix" with a fat TODO next to it for the next version.

**Comment by edmundvermeulen** [comment]
Perhaps. It is a bit of a bad experience to see memory leaks just by including the header, though.

**Comment by onqtam** [comment]
So... Could you do this fix locally for your copy of doctest for now, and wait for me to try to figure this thing out when I have the time?

**Comment by edmundvermeulen** [comment]
Yes. I have already fixed it here locally.

**Comment by onqtam** [comment]
So I just tried copy-pasting your example and am unable to reproduce this with VS 2017.

Are you running the executable in some special way? I haven't investigated what ```_CrtSetDbgFlag``` and those flags do but I assume the program would check itself upon exiting...? I tried the Debug version since in Release that call for setting flags is a no-op.

How did you diagnose that the problem was the thread-local global? There are other thread-local global variables apart from it...

**Comment by edmundvermeulen** [comment]
This is with a clean new Console Application project created via the project wizard.

Don't you see the memory leaks in Visual Studio's output window?

**Comment by edmundvermeulen** [comment]
To be clear: The memory leaks are not printed with the console output, but rather in the Output window inside Visual Studio itself.

**Comment by onqtam** [comment]
I previously tried it in a solution generated with CMake but now just tried it with a new Console Application project created from the studio - and got no errors. I used VS 2019 Preview 4.3.
![no_problem](https://user-images.githubusercontent.com/4648169/54866405-f77d2200-4d7b-11e9-9e4d-6c2ab5c1be25.png)

I highly suspect that this has nothing to do with doctest... memory leaks are diagnosed on the CI with industry-standard tools such as valgrind and the sanitizers on Linux. And it is really strange that only that thread-local global is problematic.

I'm currently out of ideas about this and don't want to commit fixes which I don't understand. I just released version 2.3.0 but will keep this issue open if any new information comes up - for now I'm clueless :|

**Comment by onqtam** [comment]
Well I'm going to close this... It will remain for others to find and inspect if anyone else hits this problem, but as far as I understand there shouldn't be anything wrong with that thread local, and there are no problems with the other thread locals as well... I cannot reproduce this locally - I'll need even more information on how to reproduce it - with an even smaller example... perhaps using cl.exe on the command line with the full list of flags or something, and not calling the executable through the debugger maybe... I don't know. But the fix suggested here doesn't seem right to me. You could still define DOCTEST_THREAD_LOCAL to nothing locally.

**Comment by drorgl** [comment]
A related issue might show up when using doctest with mingw32/64.

The simplest example would be:
```
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include <doctest/doctest.h>

TEST_CASE("simple test")
{
    CHECK(1 == 1);
}
```

run with "-s" to see the issue. when debugging with gdb it shows a segfault.

Adding `#define DOCTEST_THREAD_LOCAL` removes the segfault but I did not run it through memory analyzer to see if there is a leak as well.

**Comment by otrempe** [comment]
We started using doctest and came across this issue.

The snippet provided in the original message did not reproduce the leak detection output.
 
This snippet should reproduce it

```
#include <crtdbg.h>

#define DOCTEST_CONFIG_IMPLEMENT
#include "doctest.h"

int main(int argc, char** argv)
{
    const auto nFlags = _CrtSetDbgFlag(_CRTDBG_REPORT_FLAG);
    _CrtSetDbgFlag(nFlags | _CRTDBG_ALLOC_MEM_DF | _CRTDBG_LEAK_CHECK_DF);
    _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_DEBUG);

	_CrtDumpMemoryLeaks();

    return 0;
}
```

**Comment by otrempe** [comment]
@mitchgrout You seem like the maintainer of this library.

Tagging you just to make sure you see the last comment I added to this issue since it's closed and you were not part of the original discussion.

**Comment by mitchgrout** [comment]
Unfortunately, I'm not able to reprod this issue with your example on my setup (Windows 11, VS Community 2026, cl-19.51.36256, x64 build). If you're able to look into exactly what might be leaking, I'm happy to get it patched, but I won't be able do much it seems.

**Comment by otrempe** [comment]
The snippet has to be launched in debug configuration with the debugger attached (F5) to produce a leak report in the output window.
Did you try that?

I'll give some more context about the issue.

First, there are no real leaks. These are just apparent leak detection. They are caused by _CrtDumpMemoryLeaks(); being called before the static-duration objects of doctest being cleaned up at program termination.

In the above snippet I have control over the call site of _CrtDumpMemoryLeaks(). But it is not always the case.

Even though it is being less used for newer applications, MFC is still common for Windows desktop application and legacy applications.
It is a known behavior that MFC calls _CrtDumpMemoryLeaks() when unloaded. This causes the detection of apparent memory leaks for components that have not yet been cleaned up. 

I think it would be great to adapt doctest so that memory leaks are not falsely reported when it is used in the context of an application using MFC.

I made some local changes to doctest to see how it could be implemented, and it turned out to be pretty minimal. I could clean this up and submit a PR if it is ok with you.

**Comment by mitchgrout** [comment]
> The snippet has to be launched in debug configuration with the debugger attached (F5) to produce a leak report in the output window. Did you try that?

Ah yeah, missed that; yep, that gives reprod for me.

> First, there are no real leaks. These are just apparent leak detection. They are caused by _CrtDumpMemoryLeaks(); being called before the static-duration objects of doctest being cleaned up at program termination.

There's a few function-static objects around. My guess is that it's objecting mostly to the listener / reporter / exception-translator collections we have? If the fix is relatively straightforward I'm all for it.
